### ISSUE TITLE
Array of elements passed to Typeset is ignored if no callback is given

### ISSUE THREAD

**Issue by jdleesmiller** [mathjax/MathJax#1680]
The documentation for `Hub.Typeset` suggests that you can pass an array of elements or a callback or both:
![image](https://cloud.githubusercontent.com/assets/160829/21453147/4256d4c8-c904-11e6-9e42-2eb539434b77.png)

However, the argument handling in the `elementCallback` [helper method](https://github.com/mathjax/MathJax/blob/6273842a9746731b9ecca0de18ec9fd50a36df99/unpacked/MathJax.js#L2427-L2428) has some (IMHO) odd behaviour when `element` is an array and `callback` is not specified. The code in question is:
```js
if (callback == null && (MathJax.Object.isArray(element) || typeof element === 'function'))
  {try {MathJax.Callback(element); callback = element; element = null} catch(e) {}}
```

The result is that `callback` ends up being set to the first element in the array, and `element` is set to `null`. The `element` then defaults to the document's `body` later on, which in most cases masks the bug (unless the intended elements aren't yet in the`body`, in which case much confusion ensues). However, if you use `signal.Interest` to track the events, you can see that the element is not actually tracked.

It's also worth noting that this affects other hub functions, such as `Update`.

### Example

Run this in the dev console on the home page:
```js
MathJax.Hub.signal.Interest(function (message) {console.log(message)});
MathJax.Hub.Queue(['Typeset', MathJax.Hub, $('.hero-math').toArray()]);
MathJax.Hub.Queue(['Typeset', MathJax.Hub, $('.hero-math').toArray(), function callbackHere() {}])
```

Output:
![image](https://cloud.githubusercontent.com/assets/160829/21453315/a6930b22-c905-11e6-826c-872d8ef0b0c5.png)

When the callback is not specified, the signal messages indicate that MathJax is searching the `body`. When the callback is specified, the signal messages indicate that MathJax is searching the `div.hero-math` element, which is what was intended.

**Comment by dpvc** [comment]
The issue here is a little more complicated than it might appear, and has to do with the fact that a callback (in MathJax) need not be a function, but can be specified in a variety of other ways (see [the callback documentation](http://docs.mathjax.org/en/latest/api/callback.html)).

When a single argument is passed to `Typeset()`, MathJax needs to decide whether it is a callback or an element (either an id of one, or the DOM element itself, or an array of those), and this is a bit tricky.  The line you quote above is intended to do that (but it isn't perfect).  

``` javascript
if (callback == null && (MathJax.Object.isArray(element) || typeof element === 'function'))
  {try {MathJax.Callback(element); callback = element; element = null} catch(e) {}}
```

**Comment by dpvc** [comment]
The issue here is a little more complicated than it might appear, and has to do with the fact that a callback (in MathJax) need not be a function, but can be specified in a variety of other ways (see [the callback documentation](http://docs.mathjax.org/en/latest/api/callback.html)).

When a single argument is passed to `Typeset()`, MathJax needs to decide whether it is a callback or an element (either an id of one, or the DOM element itself, or an array of those), and this is a bit tricky.  The line you quote above is intended to do that (but it isn't perfect).  

``` javascript
if (callback == null && (MathJax.Object.isArray(element) || typeof element === 'function'))
  {try {MathJax.Callback(element); callback = element; element = null} catch(e) {}}
```

Here, if there is no callback (i.e., only one argument was passed), and if that argument looks like it could be a callback specification (a callback can be a function or an array with the right kind of elements), then `Typeset()` tries to create a callback out of the argument, and if it succeeds, it uses the argument as its `callback` and sets the `element` to `null` (otherwise it keeps the argument as the `element` and leaves the `callback` as `null`.  Because the array form of a callback is complicated, `Typeset()` leave the checking for a valid callback to `MathJax.Callback()`.  (Technically, a callback can also be specified as an object with certain properties, so the code here misses that, but it is a rare usage.)

It turns out that the situation that causes problems is when the single argument to `Typeset()` is an array of only one element.  If `$('.hero-math').toArray()` returned more than one element, everything would be OK, as I describe below.

The problem actually is in `MathJax.Callback()`, not in `MathJax.Hub.Typeset()`.  One of the ways of specifying a callback is as a function, but MathJax also accepts an array consisting of a single function as a specification for a callback.  As it tries to determine the format of a callback, `MathJax.Callback()` "unpacks" a singleton-array into its single element at [line 301](https://github.com/mathjax/MathJax/blob/master/unpacked/MathJax.js#L301), and this is the source of the problem.  If the array consists of a single DOM element, `MathJax.Callback()` turns it into that single DOM element, and then thinks that this is giving the callback as an object literal (one of the valid forms), so a callback is created.  (If it is an array of two DOM elements, then it is not unpacked, and the array ends up not matching a valid callback, and none is returned, as expected.)

So line 301 is the culprit.  It should only unpack the array if it is an array whose one element is a *function* object, not just *any* object.

Until that is fixed, it is best to use

    MathJax.Hub.Queue(['Typeset', MathJax.Hub, $('.hero-math').toArray(), {}]);

to force there to be an empty callback if the query select could return an array with only one element.

**Comment by dpvc** [comment]
The `issue1680` branch of my fork of MathJax includes a fix for this.

**Comment by dpvc** [comment]
==> Merged

**Comment by dpvc** [comment]
==> In testsuite

API/Callback/issue1680.html
